### ISSUE TITLE
Is  `std::fs::copy` racy and/or under-specified in how it sets file permissions?

### ISSUE THREAD

**Issue by briansmith** [rust-lang/rust#93126]
The documentation for `std::fs::copy` says "This function will also copy the permission bits of the original file to the destination file."

I do think the function adheres to that contract. But, is that contract strong enough? 

Consider a file that is isn't readable by the group or world that is passed to `std::fs::copy`. My expectation is that the copied file would never have group or world file permissions different from the source file. However, currently, between the time `fs::File::create` is called and the time `writer.set_permissions` is called, the destination file may have group/world permissions different than (and in particular, potentially more liberal than) the source file's permissions.

Further, what if, for some reason, the `set_permissions` call fails, or the process is killed before then? Then with the current code we'd be left with a group-/world-readable file. 

It seems like instead, when we create the file, we should use a syscall that ensures the file is created with the source file's permissions. 

If the file is read-only for the current user, then maybe we need to add the user write permission during creation, and then remove it when we're done writing? If we do this, then we should document that the file may be left writable if errors occur or the process crashes.

At a minimum, the documentation needs to be updated to consider these cases.

```
pub fn copy(from: &Path, to: &Path) -> io::Result<u64> {
    let mut reader = fs::File::open(from)?;
    let metadata = reader.metadata()?;

    if !metadata.is_file() {
        return Err(NOT_FILE_ERROR);
    }

    let mut writer = fs::File::create(to)?;
    let perm = metadata.permissions();

    let ret = io::copy(&mut reader, &mut writer)?;
    writer.set_permissions(perm)?;
    Ok(ret)
}
```

**Comment by the8472** [comment]
> If the file is read-only for the current user, then maybe we need to add the user write permission during creation, and then remove it when we're done writing?

At least on linux this isn't necessary. The current file descriptor permissions and the create-mode are independent.
Additionally we should use `O_TMPFILE` on platforms that support it and only link the file in its directory entry once the copying has finished, that'll avoid leaving any dead files behind if we run out of space or something like that.

**Comment by briansmith** [comment]
> > If the file is read-only for the current user, then maybe we need to add the user write permission during creation, and then remove it when we're done writing?
> 
> At least on linux this isn't necessary.

Right, POSIX says "The third argument does not affect whether the file is open for reading, writing, or for both." I'm not sure about non-POSIX filesystems though.

**Comment by briansmith** [comment]
> Additionally we should use `O_TMPFILE` on platforms that support it and only link the file in its directory entry once the copying has finished, that'll avoid leaving any dead files behind if we run out of space or something like that.

The permissions temporarily (or, in the case of a failure to set the permissions) is a security issue in general, as the contents of the source file are (at least temporarily) exposed beyond what's intended.

Although in some cases writing a partial file can also be a security issue (e.g. a truncated `/etc/sudoers` could be a very bad thing), in general we can't protect against all of the issues that might result in the file only being partially written, at least without `fdatasync()` or `fsync()`, and so maybe that should be out of scope.
